File: Python/backtracking/n_queens.py
"""
n_queens.py
------------
Python solution for the N-Queens Problem using backtracking.

Description:
- The program places N queens on an N x N chessboard so that no two queens attack each other.
- It prints all possible solutions and the total number of solutions.

Usage:
    python n_queens.py
"""

def print_board(board):
    """Print the chessboard configuration."""
    for row in board:
        print(" ".join(row))
    print("\n")


def is_safe(board, row, col, n):
    """Check if placing a queen at (row, col) is safe."""
    # Check this column
    for i in range(row):
        if board[i][col] == "Q":
            return False

    # Check upper-left diagonal
    i, j = row - 1, col - 1
    while i >= 0 and j >= 0:
        if board[i][j] == "Q":
            return False
        i -= 1
        j -= 1

    # Check upper-right diagonal
    i, j = row - 1, col + 1
    while i >= 0 and j < n:
        if board[i][j] == "Q":
            return False
        i -= 1
        j += 1

    return True


def solve_n_queens(board, row, n, solutions):
    """Backtracking function to solve N-Queens problem."""
    if row == n:
        # Found a solution, add a copy to solutions list
        solution = ["".join(r) for r in board]
        solutions.append(solution)
        print_board(board)
        return

    for col in range(n):
        if is_safe(board, row, col, n):
            board[row][col] = "Q"
            solve_n_queens(board, row + 1, n, solutions)
            board[row][col] = "."  # backtrack


def main():
    n = int(input("Enter the number of queens (N): "))
    board = [["." for _ in range(n)] for _ in range(n)]
    solutions = []
    print(f"\nAll possible solutions for {n}-Queens problem:\n")
    solve_n_queens(board, 0, n, solutions)
    print(f"Total solutions found: {len(solutions)}")


if __name__ == "__main__":
    main()

File: CPP/Backtracking/N_queens.cpp
/*
Problem statement : 

The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.

Given an integer n, return all distinct solutions to the n-queens puzzle. You may return the answer in any order.

Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.' both indicate a queen and an empty space, respectively.
  Input: n = 4
Output: [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
Explanation: There exist two distinct solutions to the 4-queens puzzle as shown above
Example 2:

Input: n = 1
Output: [["Q"]]
 

Constraints:

1 <= n <= 9
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // check if placing queen at (row,col) is safe
    bool isSafe(vector<vector<int>>& board, int row, int col) {
        int x = row, y = col;
        while (y >= 0) if (board[x][y--] == 1) return false;
        x = row; y = col;
        while (x < board.size() && y >= 0) if (board[x++][y--] == 1) return false;
        x = row; y = col;
        while (x >= 0 && y >= 0) if (board[x--][y--] == 1) return false;
        return true;
    }

    // convert board to string format
    void addSol(vector<vector<int>>& board, int n, vector<vector<string>>& ans) {
        vector<string> arr;
        for (int i = 0; i < n; i++) {
            string str = "";
            for (int j = 0; j < n; j++)
                str.push_back(board[i][j] ? 'Q' : '.');
            arr.push_back(str);
        }
        ans.push_back(arr);
    }

    // backtrack to find all solutions
    void solve(vector<vector<int>>& board, int n, int col, vector<vector<string>>& ans) {
        if (col == n) { addSol(board, n, ans); return; }
        for (int row = 0; row < n; row++) {
            if (isSafe(board, row, col)) {
                board[row][col] = 1;
                solve(board, n, col + 1, ans);
                board[row][col] = 0;
            }
        }
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<vector<int>> board(n, vector<int>(n, 0));
        vector<vector<string>> ans;
        solve(board, n, 0, ans);
        return ans;
    }
};

int main() {
    int n;
    cout << "Enter number of queens: ";
    cin >> n;
    Solution s;
    auto res = s.solveNQueens(n);
    cout << "Total Solutions: " << res.size() << "\n\n";
    for (auto& sol : res) {
        for (auto& row : sol)
            cout << row << "\n";
        cout << "\n";
    }
    return 0;
}
